## 异步编程

异步编程的主要方式：

* 回调函数
* 发布/订阅
* Promise
* Generator
* Async/Await

### 回调函数

回调函数是最简单的异步编程方式，也是 Node.js 异步 I/O 最核心的实现方式。

#### 基本特点

回调函数的优点：

- 足够简单，容易理解和部署

回调函数的缺点：

- 回调函数内的异常难以追踪和捕获，比较 示例1 和 示例2，示例2 中在回调中抛出错误时，其同步调用栈已不存在，故无法追踪完整的调用信息。一种常见的策略就是 Node.js 中使用的错误优先的回调，即将错误作为第一个参数传递给回调函数，这也是为什么 Node.js 约定回调函数作为第一个参数的原因。

  示例1：

  ![img](/Users/flying/Documents/Program/%E7%AC%94%E8%AE%B0/WEB/images/9607.png)

  ```
  function a() {
      b();
  }
  
  function b() {
      c();
  }
  
  function c() {
      d();
  }
  
  function d() {
      throw new Error('出错啦');
  }
  
  a();
  ```

  示例2：

  ![img](/Users/flying/Documents/Program/%E7%AC%94%E8%AE%B0/WEB/images/9608.png)

  ```
  function a() {
      b();
  }
  
  function b() {
      c(cb);
  }
  
  function c(callback) {
      setTimeout(callback, 0)
  }
  
  function cb() {
      throw new Error('出错啦');
  }
  
  a();
  ```

- 高度耦合，难以维护，比如回调地狱，多重异步操作相互依赖并需先后顺次执行

  ```
  step1(function (value1) {
      step2(function (value2) {
          step3(function (value3) {
              step4(function (value4) {
                  // Do something with value4
              });
          });
      });
  });
  ```

  可以通过将嵌套操作封装成独立函数的方式，解锁回调地狱

  ```
  function fun1(params) {
      // todo
      asyncFunction(fun2);
  }
  
  function fun2(params) {
      // todo
      asyncFunction(fun3)
  }
  
  function fun3(params) {
      // todo
      asyncFunction(fun4)
  }
  
  function fun4(params) {
      // todo
  }
  
  asyncFunction(fun1)
  ```

- 协同困难，比如

  - 需要同时获取多个异步 API 的结果后才能执行下一步，例如，需要获得多个接口数据后渲染页面
  - 需要获取多个异步 API 中最先完成的结果后执行下一步，例如，考虑足够的性能，同时从 Cache Storage 和网络获取同一个资源，哪个先拿到用哪个

#### 应用实践

浏览器中的回调函数：

```
asyncFunction(function(value){
		// todo
})
```

Node.js 中的回调函数：

```
fs.readFile('file.js', (err, data) => {
    if (err) {
        throw err;
    }

    // to do something with data
});
```

### 发布/订阅

发布/订阅模式，也称为观察者模式，它定义了一种对象之间一对多的依赖关系，当一个对象的状态发生改变时，所有依赖它的对象都将得到通知。它由两类对象组成：

* 发布者，负责发布事件，而无需关注哪些模块订阅了事件
* 订阅者，通过订阅事件来观察发布者

发布/订阅模式中，发布者和订阅者是完全解耦的，彼此不知道对方的存在，两者仅共享一个自定义事件的名称。

在 JS 中，一般使用事件模型替代传统的发布-订阅模式。因为在 JS 中，一切皆对象，无需像静态类型语言那样，必需创建合乎条件的对象，然后，往往还需要向上转型才能通过严苛的接口类型检测。总之，创建发布者或订阅者对象在 JS 中显得过于刻意了，事件模型足矣。

当然，一定要严格按照发布/订阅实现的话，可以参考 Vue.js 的 Watcher() 和 Dep() 的实现。

#### 基本特点

发布/订阅模式的优点：

* 发布者和订阅者的关系解耦，便于维护
* 侦听器作为回调函数可以随意的添加或删除，这样可以自由的添加业务逻辑，并将业务逻辑有效的分割到多个函数中更好地隔离和维护

发布/订阅模式的缺点：

* 事件机制的问题 - 事件触发必需在事件监听之后，可以保存已触发事件在订阅时即行触发来解决这个问题
* 模式本身的问题 - 过度使用会导致对象与对象之间的必要联系被深埋在背后，难以追踪

#### 事件监听

相应对象（订阅者）事先绑定事件，当事件发生时，程序内部（发布者）通知并触发相应事件。

浏览器中的事件监听：

```
document.body.addEventListener('click', function () {
    alert('Click');
}, false);
```

Node.js 中的事件监听：

```
let http = require('http');

http.createServer(function (req, res) {
    let postData = '';
    
    req.setEncoding('utf8');

    // 监听请求的data事件 
    req.on('data', function (trunk) {
        postData += trunk;
    });

    // 监听请求的end事件
    req.on('end', function () {
        res.end(postData);
    });
}).listen(8080);
```

#### 自定义事件

DOM 事件，或者 Node.js 中默认支持的事件无法满足所有的场景需求，所以，需要自定义事件系统。

浏览器中的自定义事件原型如下。当需要某个对象拥有事件能力时，可以将 Event 对象的属性 extend 到相应的对象，类似于 Backbone。

```
let Event = {
    clientList: {},
    listen(name, callback) {
        let self = this;

        if (!self.clientList[name]) {
            self.clientList[name] = [];
        }

        self.clientList[name].push(callback);
    },
    remove(name, callback) {
        let self = this;
        let items = self.clientList[name];

        if (!items || items.length <= 0) {  // 没有事件
            return false;
        }

        if (!callback) {                    // 清除某个事件的所有回调
            items = [];
        } else {                            // 清除某个事件的单个回调
            for (let i = items.length; --i;) {
                if (items[i] === callback) {
                    items.splice(i, 1);
                }
            }
        }
    },
    trigger(name) {
        let self = this;
        let args = [].slice.call(arguments, 1);
        let items = self.clientList[name];

        if (!items || items.length <= 0) {  // 没有事件
            return false;
        }

        items.forEach(function (fn) {
            fn.apply(self, args);
        });
    }
};
```

Node.js 中的自定义事件：

Nodejs 中通过 events 模块实现了原生的发布/订阅模式。

```
let Emitter = require('events');
let emitter = new Emitter();

emitter.on('someEvent', function (stream) {
    console.log(stream + 'from eventHandler1');
});

emitter.on('someEvent', function (stream) {
    console.log(stream + 'from eventHandler2');
});

emitter.emit('someEvent', 'I am a stream!');
```

### Promise

Promise 相当于异步操作的占位符，它不会订阅一个事件，也不会传递一个回调函数给目标函数，而是让函数返回一个异步操作的 Promise，例如

```
// readFile承诺在未来的某个时刻完成
let promise = readFile('file.js');
```

这样，程序不会阻塞，至于未来对这个对象的操作，完全取决于 Promise 的生命周期。

Promise 的生命周期很简单，只有三种状态：

- pending，Promise 进行中
- fulfilled，Promise 成功完成
- rejected，Promise 异步操作失败，可能是程序错误或其他的原因

Promise 的状态规则：

- 当前状态存储在内在属性 [[PromiseState]] 中，不能读取 
- 状态只能从 pending 转变为 fulfilled 或 rejected，且不能回退
- 状态改变后通过 then 或者 catch 添加的操作函数依旧会被触发

创建 Promise 的方法：

- Promise 构造函数，创建未完成的 Promise 对象
- Promise.resolve()，创建成功完成态的 Promise 对象
- Promise.reject()，创建失败完成态的 Promise 对象

响应多个 Promise 的方法：

- Promise.all(_Promise 对象数组_)，所有 Promise 都完成后执行后续 then/catch
- Promise.race(_Promise 对象数组_)，只要有一个 Promise 完成即执行后续 then/catch

Promise 的使用注意：

- 每次调用 then() 或者 catch() 方法，其实都会创建并返回一个新的 Promise 对象，所以 Promise 才可以链式调用。以下两种写法是等效的。

  ```
  // 写法1
  let p1 = new Promise(function (resolve, reject) {
      resolve(1);
  });
  
  p1.then(function (value) {
      console.log(value);
  }).then(function () {
      console.log('Finished');
  });
  
  // 写法2
  let p1 = new Promise(function (resolve, reject) {
      resolve(1);
  });
  
  let p2 = p1.then(function (value) {
      console.log(value);
  });
  
  p2.then(function () {
      console.log('Finished');
  });
  ```

- Promise.resolve() 和 Promise.reject() 方法的参数如果是一个 Promise 对象时，则直接返回参数 Promise 对象；如果参数是一个 Thenable 对象，则  Thenable 对象会被包装成一个 Promise 对象返回。

  > Thenable 对象：拥有接受 resolve 和 reject 两个参数函数的 then() 方法的普通对象

  ```
  let thenable = {
      then(resolve, reject) {
          resolve(42);
      }
  };
  
  let p1 = Promise.resolve(thenable);
  
  p1.then(value => {
      console.log(value);    // 42
  });
  ```

- Promise 链中抛出的错误具有“冒泡”性质，会一直向后传递，直至被最近的 catch 捕获

  ```
  let p1 = new Promise(function (resolve, reject) {
      throw new Error('Error A');
  });
  
  p1.catch((err) => {
      console.log(err.message);			// Error A
      throw new Error('Error B');
  }).catch((err) => {
      console.log(err.message);			// Error B
  });
  ```

- 在 Promise 链中可以通过 return 传递数据

  - return 为非 Promise 对象值时，返回值传递给下一个 then 函数

    ```
    let p1 = new Promise(function (resolve, reject) {
        resolve(1);
    });
    
    p1.then((value) => {
        console.log(value);		// 1
    
        return value + 1;
    }).then((value) => {
        console.log(value);		// 2
    });
    ```

  - return 为 Promise 对象时，返回一个新的相应状态的 Promise 对象

    ```
    // 写法1
    let p1 = new Promise(function (resolve, reject) {
        resolve(1);
    });
    
    let p2 = new Promise(function (resolve, reject) {
        resolve(2);
    });
    
    p1.then((value) => {
        console.log(value);		// 1
        return p2;
    }).then((value) => {
        console.log(value);		// 2
    });
    
    // 写法2
    let p1 = new Promise(function (resolve, reject) {
        resolve(1);
    });
    
    let p2 = new Promise(function (resolve, reject) {
        resolve(2);
    });
    
    let p3 = p1.then((value) => {
        console.log(value);		// 1
        return p2;
    });
    
    console.log(p2 === p3);		// false
    
    p3.then((value) => {
        console.log(value);		// 2
    });
    
    ```

#### 基本特点

Promise 并没有从根本上消除回调函数，只是拉平，或者说通过 then()/catch() 函数延迟了回调函数的绑定。

Promise 的优点：

- 链式调用更符合编程习惯
- 与回调函数存在协同问题不同，Promise.all() 和 Promise.race() 让对象协同更加方便
- 与 Generator 和 await 实现异步编程一定程度上阻塞了进程不同，Promise 不存在这样的问题

Promise 的缺点：

- 需要创建额外的 Promise 对象

- 不能够中止 promise 链，例如：

  ```
  let promise = new Promise((resolve) => {
      setTimeout(() => {
          resolve('hello');
      }, 1000);
  });
  
  promise.then((value) => {
      throw new Error('Error');
  }).then(value => {
      console.log(value);
  }, error => {
      console.log(error.message);
  }).then(function () {
      console.log('Continue');
  });
  
  // Error
  // Continue
  ```

- 如果不设置 catch 函数，Promise 内部的错误不会反映到外部。不过，最新的浏览器和 Node.js 已经意识到了这一点：

  - 它们都会在全局上下文提供 unhandledRejection 事件用于处理程序未主动 catch 的错误
  - Chrome 和 Firefox 浏览器的测试发现，如果为 catch 错误，浏览器会直接报错，并中断 JS 的执行

#### 原理实现

Promise 原理实现代码

```
class RealPromise {
    constructor() {
        this.isPromise = true;
        this.handleQueue = [];
    }

    then(resolve, reject) {
        let self = this;
        let handle = {};

        if (typeof resolve === 'function') {
            handle.resolve = resolve;
        }

        if (typeof reject === 'function') {
            handle.reject = reject;
        }

        self.handleQueue.push(handle);

        return self;
    }

    catch(reject) {
        let self = this;
        let handle = {};

        if (typeof reject === 'function') {
            handle.reject = reject;
        }

        self.handleQueue.push(handle);

        return self;
    }
}

class Deferred {
    constructor() {
        this.state = 'pending';
        this.promise = new RealPromise();
    }

    resolve(obj) {
        let handle;
        let handleQueue = this.promise.handleQueue;

        this.state = 'fufilled';

        while (handle = handleQueue.shift()) {

            if (handle.resolve) {
                let res = handle.resolve(obj);

                if (typeof res === 'object' && res.isPromise) {
                    res.handleQueue = handleQueue;
                    this.promise = res;
                }

                return;
            }
        }
    }

    reject(obj) {
        let handle;
        let handleQueue = this.promise.handleQueue;

        this.state = 'rejected';

        while (handle = handleQueue.shift()) {

            if (handle.reject) {
                let res = handle.reject(obj);

                if (typeof res === 'object' && res.isPromise) {
                    res.handleQueue = handleQueue;
                    this.promise = res;
                }

                return;
            }
        }
    }
}

class Promise {
    constructor(callback) {
        let defer = new Deferred();

        callback(defer.resolve.bind(defer), defer.reject.bind(defer));

        return defer.promise;
    }

    static resolved() {
    }

    static rejected() {
    }
}
```

#### 应用实践

node.js 中 api 的 promise 化

```
function promisify(fn) {
    return function () {
        let args = [].slice.call(arguments);

        return new Promise(function (resolve, reject) {
            fn.apply(null, args.push(function (err, ...args) {
                if (err) {
                    return reject(err);
                }

                resolve(args);
            }))
        });
    }
}
```

### Generator

#### 迭代器

> iterator /ɪtə'reɪtə/ n. 迭代器, 迭代程序

传统迭代必需初始化一个记录迭代位置的变量，同时，多层嵌套迭代时，维护位置变量复杂且易出错。所以，ES6 参考其他语言，增加了程序化的迭代方式 - 迭代器，来简化迭代操作，并减少循环中出错的几率。

迭代器，iterator，是一种特殊的对象，其具有专门为迭代过程设计的 API。迭代器对象有一个 next 方法，调用该方法返回一个对象，该对象具有两个属性：

* value，当前迭代位置的值
* done，布尔值，标识迭代是否完成

使用 ES5 的方式设计一个简单的生成迭代器的函数

```
// set 为集合类型
function createIterato(set) {
    let i = 0;
    
    return {
        next() {
            let done = i >= set.length;

            return {
                done: done,
                value: done ? undefined : set[i++]
            }
        }
    }
}
```

#### 生成器

为了创建迭代器更加方便，ES6 添加了生成器。

生成器是一种返回迭代器的特殊函数，通过在 function 关键字和函数名之间添加雪花号*来标识生成器。在这种函数中，使用 yield 关键字标识每次迭代的位置。

```
// 生成器
function *createIterator() {
    yield 1;
    yield 2;
    yield 3;
}

// 迭代器
let iterator = createIterator();

console.info(iterator.next());      // {value: 1, done: false}
console.info(iterator.next());      // {value: 2, done: false}
console.info(iterator.next());      // {value: 3, done: false}
console.info(iterator.next());      // {value: undefined, done: true}
```

生成器的使用注意：

- 不能用箭头函数创建，为啥？

- 可迭代对象 (String、Array、Set、Map、DOM 中的 NodeList 类型) 具有的 Symbol.iterator 属性本质上也是生成器，使用展开运算符或者 for..of 时，引擎内部就是调用的该属性。

  ```
  let divs = document.querySelectorAll('div');
  
  for (let div of divs) {
      console.log(div.id);
  }
  ```

  所以，可以通过添加 Symbol.iterator 生成器属性的方式使得一个普通对象成为可使用 for..of 方法的可迭代对象

  ```
  let colletion = {
      items: [1, 2, 3],
      * [Symbol.iterator]() {
          for (let item of this.items) {
              console.log(item);
          }
      }
  }
  
  for (let value of  colletion) {
      console.log(value);
  }
  ```

- 生成器创建的迭代器对象还有一个 throw 方法，用来主动抛出错误

  - 如果抛出的错误得不到捕捉， JS 引擎报错，进程结束
  - 如果抛出的错误得到捕捉，程序继续执行

  ```
  // 生成器
  function* createIterator() {
      yield 1;
  
      try {
          yield 2;
      } catch (e) {
          yield 3;
      }
  
      yield 4;
  }
  
  // 迭代器
  let iterator = createIterator();
  let error = new Error('Boom');
  
  console.info(iterator.next());          // {value: 1, done: false}
  console.info(iterator.next());          // {value: 2, done: false}
  console.info(iterator.throw(error));    // {value: 3, done: false}
  console.info(iterator.next());          // {value: 4, done: false}
  console.info(iterator.next());          // {value: undefined, done: true}
  ```

- 生成器可以使用 return 提前返回，不过 next() 方法会接受 return 返回的值，展开运算符和 for…of 不会

  ```
  // 生成器
  function* createIterator() {
      yield 1;
      return 2;
      yield 3;
  }
  
  // 迭代器
  let iterator = createIterator();
  
  console.info(iterator.next());      // {value: 1, done: false}
  console.info(iterator.next());      // {value: 2, done: true}
  console.info(iterator.next());      // {value: undefined, done: true}
  console.info(iterator.next());      // {value: undefined, done: true}
  
  console.info(...createIterator());  // 只输出 1
  
  for (let num of createIterator()) {
      console.log(num);               // 只输出 1
  }
  ```

- 生成器的执行可以委托给其他生成器执行，从而形成生成器的嵌套

  ```
  // 生成器
  function* createIteratorNum() {
      yield 1;		// 暂停位置1
      yield 2;		// 暂停位置1
      return 3;
  }
  
  function* createIterator() {
  		// 这里的yield后带有*，表示是委托生成器，程序不会在这里暂停
      let result = yield* createIteratorNum();
      yield result;		// 暂停位置3
  }
  
  // 迭代器
  let iterator = createIterator();
  
  console.info(iterator.next());      // {value: 1, done: false}
  console.info(iterator.next());      // {value: 2, done: false}
  console.info(iterator.next());      // {value: 3, done: false}
  console.info(iterator.next());      // {value: undefined, done: true}
  ```

  对于委托生成器，需要注意的是：

  * yield 关键字后跟*是委托生成器的标志，迭代不会在这里暂停，这是 yield 关键字的第二个用法
  * 被委托方 return 返回的值会作为委托方整个 yield 表达式的值

yield 关键字的使用注意：

- 只能在生成器函数中使用

- 可以 yield 任何类型值或表达式

- yield 表达式默认返回值为 undefined，如果需要返回特定值，使用 next() 方法的参数

  ```
  // 生成器
  function* createIterator() {
      let first = yield 1;
      let second = yield first + 2;
  
      yield second + 3;
  }
  
  // 迭代器
  let iterator = createIterator();
  
  console.info(iterator.next());      // {value: 1, done: false}
  console.info(iterator.next(2));     // {value: 4, done: false}
  console.info(iterator.next(3));     // {value: 6, done: false}
  console.info(iterator.next());      // {value: undefined, done: true}
  ```

#### 异步应用

生成器最重要的应用就是异步编程，基本原理是：

* 利用程序在 yield 关键字处暂停的机制，使用 yield 执行一个异步的 API，异步 API 执行期间，程序处于暂停状态
* 异步执行成功时，调用生成器的 next() 方法继续程序的执行，同时，通过 next() 方法的参数传递异步 API 执行的结果，从而实现函数体内外的数据交换
* 异步执行出错时，可以通过 throw() 方法抛出内部错误，这个错误可以被外部的 try..catch 捕获，从而解决在回调函数中内部异常难以追踪和捕获的问题

最终，以同步编码的方式实现异步的应用。

生成器异步应用的关键就是**设计一个驱动生成器自动运行的执行器**。

##### 基本特点

生成器实现异步应用的优点：

* 基本实现了同步编码下的异步编程，从根本上消除回调函数，其语法糖 Async/Await 更是如此
* 一定程度上解决了异步任务中的 Error 不能被外部捕获的问题

生成器实现异步应用的缺点：

* 半自动的半协程方式，需要自己实现执行器
* 利用 yield 暂停执行的特点，其实是阻塞了进程，影响了性能。尽管 CO 执行器中通过并行的方式一定程度上解决的阻塞的问题，但是需要组合不同的 Promise ，比较麻烦

##### 同步执行器

不考虑异步 API，执行器基本实现

```
// 生成器
function* createIterator() {
    let first = yield 1;
    console.log(first);

    let second = yield 2;
    console.log(second);

    let third = yield 3;
    console.log(third);
}

// 执行器
function run(gen) {
    let task = gen();
    let result = task.next();

    function step() {
        console.info(result);

        if (!result.done) {
            result = task.next(result.value);
            step();
        }
    }

    step();
}

run(createIterator);

// {value: 1, done: false}
// 1
// {value: 2, done: false}
// 2
// {value: 3, done: false}
// 3
// {value: undefined, done: true}
```

##### 异步执行器

考虑异步 API，通常的异步逻辑

```
function* createIterator(set) {
		// Node.js 常见的异步操作
    // let aync1 = yield function (cb) {
    //     fs.readfile('file.js', cb);
    // };
    
    let aync1 = yield function (cb) {
        setTimeout(function () {
            cb(null, 1);
        }, 500);
    };

    console.log(aync1);

    try {
        let aync2 = yield function (cb) {
            setTimeout(function () {
                cb(new Error('Boom'));
            }, 500);
        };
    } catch (e) {
        console.log('Error');
    }

    // ... do something ...
}
```

异步执行器的实现：

```
function run(gen) {
    let iterator = gen();
    let result = iterator.next();

    function step() {
        if (!result.done) {
        		// 当yield后为一个函数时，则执行之
            if (typeof result.value === 'function') {
                result.value(function (err, data) {
                		// 出错时，使用throw抛出错误
                    if (err) {
                        result = iterator.throw(err);
                        return;
                    }
                    
                    // 执行结束使用next参数传递执行结果
                    result = iterator.next(data);
                    step();
                });
            } else {
                result = iterator.next(result.value);
                step();
            }
        }
    }

    step();
}
```

使用执行器运行生成器

```
run(createIterator);
// 1
// Error
```

##### CO 执行器

Node.js 中使用 CO 模块的基本实现同上面的异步执行器，区别只是 CO 模块要求封装异步任务的函数返回的必需是 [thunk 函数](http://www.ruanyifeng.com/blog/2015/05/thunk.html)或者 promise

异步任务生成器：

```
function* createIterator(set) {
    let aync1 = yield function () {
        return new Promise(function (resolve, reject) {
            setTimeout(function () {
                resolve(1);
            }, 500);
        });
    };

    console.log(aync1);

    try {
        let aync2 = yield function () {
            return new Promise(function (resolve, reject) {
                setTimeout(function () {
                    reject(new Error('Boom'));
                }, 500);
            });
        };
    } catch (e) {
        console.log('Error');
    }

    // ... do something ...
}
```

CO 执行器原理代码：

```
function run(gen) {
    let iterator = gen();
    let result = iterator.next();

    return new Promise(function (resolve, reject) {
        function step() {
            result.value().then(function (data) {
                result = iterator.next(data);
                step();
            }).catch(function (err) {
                result = iterator.throw(err);
            });
        }

        step();
    });
}
```

从生成器实现异步编程的原理可以知道，异步 API 执行的时候，程序实际上是阻塞的，所以，为了提高性能，CO 模块提供了并行执行的方式：

```
// 为yield指定一个promise数组
co(function* () {
    let res = yield [
        Promise.resolve(1),
        Promise.resolve(2),
        Promise.resolve(3),
    ];
    console.log(res); // => [1, 2, 3] 
}).catch(onerror);

// 为yield指定一个类数组promise对象
co(function* () {
    let res = yield {
        1: Promise.resolve(1),
        2: Promise.resolve(2),
    };
    console.log(res); // => { 1: 1, 2: 2 } 
}).catch(onerror);
```

##### 参考

* https://github.com/tj/co/blob/master/index.js
* http://www.ruanyifeng.com/blog/2015/05/co.html

### Async/Await

async 函数就是 Generator 函数的语法糖。

例如，基于 Generator 的异步实现：

```
let fs = require('fs');

let readFile = function (fileName) {
    return new Promise(function (resolve, reject) {
        fs.readFile(fileName, function (error, data) {
            if (error) reject(error);
            resolve(data);
        });
    });
};

let gen = function* () {
    let f1 = yield readFile('/etc/fstab');
    let f2 = yield readFile('/etc/shells');
    console.log(f1.toString());
    console.log(f2.toString());
};
```

使用 async 实现：

```
let asyncReadFile = async function () {
    let f1 = await readFile('/etc/fstab');
    let f2 = await readFile('/etc/shells');
    console.log(f1.toString());
    console.log(f2.toString());
};
```

比较发现：async 函数就是将 Generator 函数的 * 换成 async，将 yield 换成 await，仅此而已。

await 关键字的使用规则：

* await 命令只能用在 async 函数中

async 函数的使用注意：

* async 函数内部 return 语句返回的值，会成为 then() 方法回调函数的参数

  ```
  async function f() {
      return 'hello world';
  }
  
  f().then(v => console.log(v));		// "hello world"
  ```

* async 函数内抛出错误，会导致返回的 Promise 对象变为 reject 状态，错误会被 catch 方法捕获

  ```
  async function f() {
      throw new Error('出错了');
  }
  
  f().then(
      v => console.log(v),
      e => console.log(e.message)     // 出错了
  );
  ```

* await 命令后是 Promise 对象或者 thenable 对象时，返回该对象的结果；否则，直接返回对应的值

  ```
  async function f() {
      // 等同于 return 123;
      return await 123;
  }
  
  f().then(v => console.log(v));      // 123
  ```

* async 函数可以保留运行堆栈

  ```
  const a = () => {
      b().then(() => c());
  };
  ```

  上面代码中，b() 等待执行完成的时候，函数 a() 继续向下执行，等 b() 运行结束，a() 早就运行结束了，b() 所在的上下文环境也会消失。如果 b() 或 c() 报错，错误堆栈将不包括 a()

  ```
  const a = async () => {
      await b();
      c();
  };
  ```

  使用 async 时，b() 在等待执行完成的时候，a() 只是暂停执行，上下文环境都保存着。如果 b() 或 c() 报错，错误堆栈将包括 a()。

#### 基本特点

async 函数被认为最完美的异步编程模式，相对于 Generator 函数的改进，主要体现在以下三点：

* 内置执行器

  Generator 函数需要类似 CO 模块的执行器，async 函数自带执行器

* 更好的语义

  async 和 await，相对于星号和 yield，语义更清楚

  * async 表示函数里有异步操作
  * await 表示紧跟在后面的表达式需要等待结果

* 更广的适用性

  * co 函数库约定，yield 命令后面只能是 Thunk 函数或 Promise 对象
  * async 函数的 await 命令后面，可以跟 Promise 对象和原始类型的值（数值、字符串和布尔值，但这时会自动转换成立即 resolved 的 Promise 对象）

* 返回值是 Promise

  async 函数返回的是 Promise 对象，比 Generator 函数返回 Iterator 对象更加方便，可以直接用 then() 方法指定下一步的操作

async 函数异步应用的优点：

* await 暂停执行后恢复，执行栈及上下文依然存在

async 函数异步应用的问题：

* 同 Generator 函数类似，采用阻塞的方式实现同步编码异步执行，比较影响性能。例如，如下代码，断点可以发现，代码会在每一个 await 处暂停，直至后面的异步操作函数执行完毕

  解决方案：如果函数内的多个 await 不需要先后执行，可以并行执行

  ```
  function async1() {
      return new Promise(function (resolve, reject) {
          setTimeout(() => {
              resolve(1);
          }, 1000);
      });
  }
  
  function async2() {
      return new Promise(function (resolve, reject) {
          setTimeout(() => {
              resolve(2);
          }, 1000);
      });
  }
  
  async function run() {
      let t1 = Date.now();
      let a1 = await async1();
      let t2 = Date.now();
      let a2 = await async2();
      let t3 = Date.now();
  
      console.info([t2 - t1, t3 - t2])  // [1003, 1001]
      return [a1, a2];
  }
  
  run().then(function (data) {
      console.info(data);								// [1, 2]
  });
  ```

  改进 run 函数：

  ```
  // 方式1，await之前提前创建所有需要并行的Promise对象
  async function run() {
      let t1 = Date.now();
      let p1 = async1();
      let p2 = async2();								// 提前创建p1、p2
  
      let a1 = await p1;
      let t2 = Date.now();
      let a2 = await p2;
      let t3 = Date.now();
  
      console.info([t2 - t1, t3 - t2])  // [1005, 0]
      return [a1, a2];
  }
  
  // 方式2，使用Promise.all并行
  async function run() {
      let t1 = Date.now();
      let a1 = await Promise.all([async1(), async2()]);
      let t2 = Date.now();
  
      console.info(t2 - t1);						// 1002
      return a1;
  }
  ```

#### 应用实践

async/await 的执行顺序，await 总是需要后面的 promise 完全执行完成，才会继续执行；所在，下面代码在浏览器和 Node.js 中返回的顺序都是 3，2，1

```
async function fnA() {
    await fnB();
    console.log(1);
}

async function fnB() {
    await fnC();
    console.log(2);
}

async function fnC() {
    console.log(3);
}

fnA().then();
// 3
// 2
// 1
```

同时，需要注意程序的执行流：先从 fnA → fnB → fnC ，然后再从 fnC → fnB → fnA 返回，从而形成环形的执行流。所以，仔细领略可以发现，await 可以让之前的代码先执行，之后的代码等待后面代码执行结束后再执行。







